src/wms/src/main/java/org/geoserver/wms/map/RenderedImageMapOutputFormat.java

```diff
@@ -1277,28 +1278,38 @@ protected GridCoverage2D readCoverage(GridCoverage2DReader reader,
         
         //
         // If we need to add a collar use mosaic or if we need to blend/apply a bkg color
-        if(!(imageBounds.contains(mapRasterArea) || imageBounds.equals(mapRasterArea))||transparencyType!=Transparency.OPAQUE) {
-            Rectangle roi = imageBounds.intersection(mapRasterArea);
-            ROI[] rois = new ROI[] { new ROIShape(!roi.isEmpty() ? roi : mapRasterArea) };
+        ImageWorker iw = new ImageWorker(image);
+        Object roiCandidate = image.getProperty("roi");
+        if(!(imageBounds.contains(mapRasterArea) || imageBounds.equals(mapRasterArea)) 
+                || transparencyType != Transparency.OPAQUE
+                || iw.getNoData() != null || roiCandidate instanceof ROI) {
+            ROI roi;
+            if(roiCandidate instanceof ROI) {
+                ROI imageROI = (ROI) roiCandidate;
+                roi = imageROI.intersect(new ROIShape(mapRasterArea));
+            } else {
+                Rectangle intersection = imageBounds.intersection(mapRasterArea);
+                roi = new ROIShape(!intersection.isEmpty() ? intersection : mapRasterArea);
+            }
```

Comment:
I was tempted, but since I want to backport, decided not to. ROIGeometry general usage also got some recent fixes which needs some more widespread testing before we go and replace all ROIShape usag with ROIGeometry imho.